---
title: "Chromosome 7 TS Gene Visualization"
author: "Shirley Zhou"
date: "3/29/2020"
output: html_document
---

Load the following package needed:
1) Gviz to perform visualization
2) GenomicRanges to build object read by Gviz 
3) biomaRt to retreive coordinates by gene name 
```{r setup}
library(Gviz)
library(GenomicRanges)
library(biomaRt)
library(tidyselect)
library(dplyr)
```

Load the lists of interest:
1) list of TS from regression RF --> 109 
2) list of TS from classfication RF --> 131
3) list of TS from data mining by Jeremy
4).List of CDR(commonly deleted region) genes --> 77
5).List of experimental hits from Jeremy's screen

genome track --> hg38

```{r}
setwd("/Users/weihan/Documents/GitHub/ts_machine_learning")
```

1. list of ts from regression RF
```{r load lists}
regrf.complete <- read.table("../data/gwide_hema_regression/ranked_result.csv", sep = ",", header = TRUE)
regrf.complete <- regrf.complete[order(regrf.complete$Gene), ]
regrf <- regrf.complete %>% filter(pred_prob > 0.3)
regrf <- data.frame(regrf, rank = c(1:109))
regrf.ts <- as.vector(regrf$Gene)
```

2.list of ts from classification RF
```{r}

classrf <- read.csv("../data/gwide_hema_classification/pred_hits_CGC_bootstrap.csv",stringsAsFactors = FALSE) %>%
        dplyr::select(-"X") 
classrf.ts <- as.vector(classrf$Gene)
classrf.ts

```

3.list of TS from data mining by Jeremy
```{r}
jeremy <- read.csv("../data/96ts_work_data.csv", sep = ",", header = TRUE)
jeremy.ts <- as.vector(jeremy$Gene)
```

4.List of CDR genes
```{r}
CDR_genes <- read.csv("../data/CDR_list.csv",stringsAsFactors = FALSE,header = FALSE) 
colnames(CDR_genes) = "CDR_gene"
CDR.genes <- as.vector(CDR_genes$CDR_gene)
```
5.List of Experimental hits, ranked by combined score
```{r}
exp_hits <- read.csv("../data/z_value_calc/comb_ranking.csv",stringsAsFactors = FALSE)
str(exp_hits)
exp.hits <- exp_hits %>%
        dplyr::select("Gene","z_comb") %>%
        #top_frac(0.25) %>%
        dplyr::select("Gene")
exp.hits <- as.vector(exp.hits$Gene)

exp.hits

```

Write a function to obtain overlaps between two gene sets
```{r}
overlap <- function(list1, list2){
        overlap <- list1[(list1 %in% list2)]
        return(overlap)
}

#overlap between experimental hits and ML classification prediction
ML_class_exp_overlap <- overlap(classrf.ts,exp.hits)
ML_class_exp_overlap

```


define a function to retrieve chromosomal coordinate information from biomaRt
```{r biomaRt function}
data_retrieve <- function(gene, mart) {
  
  # input a vector of gene names
  # return a data frame containing:
  # 1) gene name: "hgnc_symbol"
  # 2) chromosome: "chromosome_name"
  # 3) start coordinate: "start_position"
  # 4) end coordinate: "end_position"
  # 5) strand: "strand"

  annotLookup <- getBM(
    mart = mart,
    attributes = c(
      "hgnc_symbol",
      "chromosome_name",
      "start_position",
      "end_position",
      "strand"
      ),
    filter = "hgnc_symbol",
    values = gene,
    uniqueRows=TRUE)
  
  #filter rows that have 1) chromosome_name == 7 and 2) no duplicates
  result <- annotLookup %>% filter(chromosome_name == 7)
  result <- result %>% filter (!duplicated(result$hgnc_symbol))
  return(result)
}

```

```{r initalize mart object}
#initialize mart object
mart.hs <- useEnsembl(biomart = "ensembl", 
                      dataset = "hsapiens_gene_ensembl")
```

```{r alternative initalization of mart using a mirror server}
#Use ensembl mirror site if emsembl not responding
mart.hs <- useEnsembl(biomart = "ensembl", 
                      dataset = "hsapiens_gene_ensembl", 
                      mirror = "uswest")
#Valid mirror options are 'www', 'uswest', 'useast', 'asia'
```

```{r Create the result dataframes from the 5 lists}
regrf.result <- data_retrieve(regrf.ts, mart.hs)
classrf.result <- data_retrieve(classrf.ts, mart.hs)
jeremy.result <- data_retrieve(jeremy.ts, mart.hs)
classandjeremy.result <- data_retrieve(classandjeremy.ts, mart.hs)
regandjeremy.result <- data_retrieve(regandjeremy.ts, mart.hs)

#create result dataframe from the experiment hits 
exp.hits.result <- data_retrieve(exp.hits,mart.hs)
#attach the combined z score
exp_hits <- dplyr::rename(exp_hits,hgnc_symbol = Gene)
exp.hits.result <- exp.hits.result %>% inner_join(exp_hits[c("hgnc_symbol","z_comb")],by = "hgnc_symbol")

#create the result dataframe from ML classification hits
classrf.hits.result <- data_retrieve(classrf.ts,mart.hs)
#attach the ML score 
classrf <- dplyr::rename(classrf,hgnc_symbol = Gene)
classrf.hits.result <- classrf.hits.result %>% inner_join(classrf,by = "hgnc_symbol")
```

```{r Overall Plot function}
#define a function to plot an overall result

overall_plot <- function(result, title) {
  
  chr <- 'chr7'
  gen <- 'hg38'
  
  data_g <- with(result, GRanges(chr, IRanges(result$start_position,
                                              result$end_position),
                                 result$strand,
                                 id=result$hgnc_symbol))
  atrack <- AnnotationTrack(data_g, name = title)
  
  gtrack <- GenomeAxisTrack()
  
  itrack <- IdeogramTrack(genome = gen, chromosome = chr)
  
  cdr_region <-data.frame(start = c(138369119, 100117325, 144345422),
                          end = c(140078203, 102627325, 148972761))

  ht <- HighlightTrack(trackList=list(atrack), 
                       start = cdr_region$start, 
                       end = cdr_region$end,
                       chromosome = chr,
                       genome = gen)
  plotTracks(list(itrack, gtrack, ht))
}

```


```{r Regresion RF overall plot}
overall_plot(regrf.result, "Regression Random Forest TS")
```


```{r Classfication RF overall plot}
overall_plot(classrf.result, "Classification Random Forest TS")
```


```{r Datamining overall plot }
overall_plot(jeremy.result, "Data Mining TS")
```


```{r CDR plot function}
#define a function to plot result zoomed in on a specified CDR

cdr_zoomin_plot <- function(result, title, cdr_num) {
  
  cdr <-data.frame(start = c(100117325, 138369119, 144345422),
                 end = c(102627325, 140078203, 148972761))
  
  cdr_start <- cdr$start[cdr_num]
  cdr_end <- cdr$end[cdr_num]
  
  result <- result %>% filter((start_position >= cdr_start & end_position <= cdr_end) |
                            (start_position <= cdr_start & end_position >= cdr_start) |
                            (start_position <= cdr_end & end_position >= cdr_end) |
                            (start_position <= cdr_start & end_position >= cdr_end))
  
  if (nrow(result) == 0) {
    statement = paste("No TS Genes in", title)
    return (statement)
  }
  
  chr <- 'chr7'
  gen <- 'hg38'
  
  data_g <- with(result, GRanges(chr, IRanges(result$start_position,
                                              result$end_position),
                                 result$strand,
                                 group = result$hgnc_symbol,
                                 id=result$hgnc_symbol))
  atrack <- AnnotationTrack(data_g, name = title, 
                            groupAnnotation="group", fontcolor.group="darkblue", 
                            shape = "box", just.group = "left")
  
  gtrack <- GenomeAxisTrack()
  
  itrack <- IdeogramTrack(genome = gen, chromosome = chr)
  
  if ((sum(result$start_position < cdr_start) + sum(result$end_position > cdr_end)) > 0) {
    ht <- HighlightTrack(trackList=list(atrack), 
                         start = cdr_start, 
                         end = cdr_end,
                         chromosome = chr,
                         genome = gen)
    plotTracks(list(itrack, gtrack, ht))
  } else {
      plotTracks(list(itrack, gtrack, atrack), from = cdr_start - 3e5, to = cdr_end)
  }
    
}
```


```{r zoom_in plot for regrf CDR1}
cdr_zoomin_plot(regrf.result, "Regression RF CDR1", 1)
```


```{r zoom_in plot for regrf CDR2}
cdr_zoomin_plot(regrf.result, "Regression RF CDR2", 2)
```


```{r zoom_in plot for regrf CDR3}
cdr_zoomin_plot(regrf.result, "Regression RF CDR3", 3)
```


```{r zoom_in plot for classrf CDR1}
cdr_zoomin_plot(classrf.result, "Classification RF CDR1", 1)
```


```{r zoom_in plot for classrf CDR2}
cdr_zoomin_plot(classrf.result, "Classification RF CDR2", 2)
```


```{r zoom_in plot for classrf CDR3}
cdr_zoomin_plot(classrf.result, "Classification RF CD3", 3)
```


```{r zoom_in plot for Data Mining CDR1}
cdr_zoomin_plot(jeremy.result, "Data Mining CDR1", 1)
```


```{r zoom_in plot for Data Mining CDR2}
cdr_zoomin_plot(jeremy.result, "Data Mining CDR2", 2)
```


```{r zoom_in plot for Data Mining CDR3}
cdr_zoomin_plot(jeremy.result, "Data Mining CDR3", 3)
```


```{r regrf rank plot overall} 

chr <- 'chr7'
gen <- 'hg38'
result <- regrf.result
regrf.result.rank <- data.frame(result, rank = 110 - regrf[order(regrf$Gene),]$rank)
  
data_g <- with(result, GRanges(chr, IRanges(result$start_position,
                                              result$end_position),
                               result$strand,
                               group=result$hgnc_symbol,
                               id=result$hgnc_symbol))
atrack <- AnnotationTrack(data_g, name = "Regression RF Overall")
                            
gtrack <- GenomeAxisTrack()
  
itrack <- IdeogramTrack(genome = gen, chromosome = chr)

rank_data <- regrf.result.rank
rank_data <- rank_data[order(rank_data$start_position),]

rank_data_d <- with(rank_data, GRanges(chr, IRanges(rank_data$start_position,
                                                    rank_data$end_position),
                                       strand = "*",
                                       rank=rank_data$rank))
                        
dTrack <- DataTrack(rank_data_d, name="Score")

cdr_region <-data.frame(start = c(138369119, 100117325, 144345422),
                        end = c(140078203, 102627325, 148972761))


ht <- HighlightTrack(trackList=list(atrack, dTrack), 
                       start = cdr_region$start, 
                       end = cdr_region$end,
                       chromosome = chr,
                       genome = gen)
plotTracks(list(itrack, gtrack, ht), type = "b")

```

```{r regrf rank plot top 20}
chr <- 'chr7'
gen <- 'hg38'

result <- regrf.result.rank %>% filter (rank >= (nrow(result)-20)) 

data_g <- with(result, GRanges(chr, IRanges(result$start_position,
                                            result$end_position),
                               result$strand,
                               group=result$hgnc_symbol,
                               id=result$hgnc_symbol))
atrack <- AnnotationTrack(data_g, name = "Regression RF Top20",
                          groupAnnotation="group", fontcolor.group="darkblue", 
                          shape = "box", just.group = "left")
  
gtrack <- GenomeAxisTrack()
  
itrack <- IdeogramTrack(genome = gen, chromosome = chr)

rank_data <- result
rank_data <- rank_data[order(rank_data$start_position),]

rank_data_d <- with(rank_data, GRanges(chr, IRanges(rank_data$start_position,
                                                    rank_data$end_position),
                                       strand = "*",
                                       rank=rank_data$rank))
                        
dTrack <- DataTrack(rank_data_d, name="Score")

cdr_region <-data.frame(start = c(138369119, 100117325, 144345422),
                        end = c(140078203, 102627325, 148972761))


ht <- HighlightTrack(trackList=list(atrack, dTrack), 
                       start = cdr_region$start, 
                       end = cdr_region$end,
                       chromosome = chr,
                       genome = gen)
plotTracks(list(itrack, gtrack, ht), type = "b")

```

```{r CDR plot with rank function}
#define a function to plot result zoomed in on a specified CDR

cdr_zoomin_plot_w_rank <- function(result, title, cdr_num) {
  
  cdr <-data.frame(start = c(100117325, 138369119, 144345422),
                   end = c(102627325, 140078203, 148972761))
  
  cdr_start <- cdr$start[cdr_num]
  cdr_end <- cdr$end[cdr_num]
  
  result <- result %>% filter((start_position >= cdr_start & end_position <= cdr_end) |
                              (start_position <= cdr_start & end_position >= cdr_start) |
                              (start_position <= cdr_end & end_position >= cdr_end) |
                              (start_position <= cdr_start & end_position >= cdr_end))
  
  if (nrow(result) == 0) {
    statement = paste("No TS Genes in", title)
    return (statement)
  }
  
  chr <- 'chr7'
  gen <- 'hg38'
  
  data_g <- with(result, GRanges(chr, IRanges(result$start_position,
                                              result$end_position),
                                 result$strand,
                                 group = result$hgnc_symbol,
                                 id=result$hgnc_symbol))
  atrack <- AnnotationTrack(data_g, name = title, 
                            groupAnnotation="group", fontcolor.group="darkblue", 
                            shape = "box", just.group = "left")
  
  gtrack <- GenomeAxisTrack()
  
  itrack <- IdeogramTrack(genome = gen, chromosome = chr)
  
  rank_data <- result
  rank_data <- rank_data[order(rank_data$start_position),]

  rank_data_d <- with(rank_data, GRanges(chr, IRanges(rank_data$start_position,
                                                      rank_data$end_position),
                                         strand = "*",
                                         rank=rank_data$rank))
  
  dtrack <- DataTrack(rank_data_d, name="Score")
  
  if ((sum(result$start_position < cdr_start) + sum(result$end_position > cdr_end)) > 0) {
    ht <- HighlightTrack(trackList=list(atrack, dtrack), 
                         start = cdr_start, 
                         end = cdr_end,
                         chromosome = chr,
                         genome = gen)
    plotTracks(list(itrack, gtrack, ht), type = "b")
  } else {
    plotTracks(list(itrack, gtrack, atrack, dtrack), from = cdr_start - 3e5, to = cdr_end, type = "b")
  }
}
    
```

```{r zoom_in plot for regrf CDR1}
cdr_zoomin_plot_w_rank(regrf.result.rank, "Regression RF CDR1 with Rank", 1)
```


```{r zoom_in plot for regrf CDR2}
cdr_zoomin_plot_w_rank(regrf.result.rank, "Regression RF CDR2 with Rank", 2)
```


```{r zoom_in plot for regrf CDR3}
cdr_zoomin_plot_w_rank(regrf.result.rank, "Regression RF CDR3 with Rank", 3)
```


```{r overlay function}
#define a function that takes 2 list and 1 overlayed list with unique entries and return an array with 1s and 2s

overlay_count <- function(overlay_list, list1, list2) {
  result <- c()
  
  #if an item in the overlayed list is in both list, then the entry will be 
  for (item in overlay_list) {
    if ((item %in% list1) & (item %in% list2)) {
      result <- append(result, 2)
    } else {
      result <- append(result, 1)
    }
  }
  return(result)
}

classandjeremy.ts <- unique(as.vector(c(classrf.ts, jeremy.result$hgnc_symbol)))
classandjeremy.count <- overlay_count(classandjeremy.ts, classrf.ts, jeremy.result$hgnc_symbol)

regandjeremy.ts <- unique(as.vector(c(regrf.ts, jeremy.result$hgnc_symbol)))
regandjeremy.count <- overlay_count(regandjeremy.ts, regrf.ts, jeremy.result$hgnc_symbol)
```


```{r Classification RF and Data mining overlayed graph}

chr <- 'chr7'
gen <- 'hg38'
result <- data.frame(classandjeremy.result, count = classandjeremy.count)
result <- result[order(result$start_position), ]

data_g <- with(result, GRanges(chr, IRanges(result$start_position,
                                            result$end_position),
                               result$strand,
                               group=result$hgnc_symbol,
                               id=result$hgnc_symbol))
atrack <- AnnotationTrack(data_g, name = "Classification RF and Data Mining overlayed")
                            
gtrack <- GenomeAxisTrack()
  
itrack <- IdeogramTrack(genome = gen, chromosome = chr)

data <- with(result, GRanges(chr, IRanges(result$start_position,
                                          result$end_position),
                                       strand = "*",
                                       count=result$count))
                        
dTrack <- DataTrack(data, name="Count")

cdr_region <-data.frame(start = c(138369119, 100117325, 144345422),
                        end = c(140078203, 102627325, 148972761))


ht <- HighlightTrack(trackList=list(atrack, dTrack), 
                       start = cdr_region$start, 
                       end = cdr_region$end,
                       chromosome = chr,
                       genome = gen)
plotTracks(list(itrack, gtrack, ht), type = "b")
```

```{r TS Genes in Both Classification and Data mining}

chr <- 'chr7'
gen <- 'hg38'
result <- data.frame(classandjeremy.result, count = classandjeremy.count)
result <- result %>% filter(count == 2)

data_g <- with(result, GRanges(chr, IRanges(result$start_position,
                                            result$end_position),
                               result$strand,
                               group=result$hgnc_symbol,
                               id=result$hgnc_symbol))
atrack <- AnnotationTrack(data_g, name = "TS Genes in Both Classification RF and Data Mining",
                          groupAnnotation="group", fontcolor.group="darkblue", 
                          shape = "box", just.group = "left")
                            
gtrack <- GenomeAxisTrack()
  
itrack <- IdeogramTrack(genome = gen, chromosome = chr)

cdr_region <-data.frame(start = c(138369119, 100117325, 144345422),
                        end = c(140078203, 102627325, 148972761))


ht <- HighlightTrack(trackList=list(atrack), 
                       start = cdr_region$start, 
                       end = cdr_region$end,
                       chromosome = chr,
                       genome = gen)
plotTracks(list(itrack, gtrack, ht), )

```

```{r Regression RF and Data mining overlayed graph}

chr <- 'chr7'
gen <- 'hg38'
result <- data.frame(regandjeremy.result, count = regandjeremy.count)
result <- result[order(result$start_position), ]

data_g <- with(result, GRanges(chr, IRanges(result$start_position,
                                            result$end_position),
                               result$strand,
                               group=result$hgnc_symbol,
                               id=result$hgnc_symbol))
atrack <- AnnotationTrack(data_g, name = "Regression RF & Data Mining Overlayed")
                            
gtrack <- GenomeAxisTrack()
  
itrack <- IdeogramTrack(genome = gen, chromosome = chr)

data <- with(result, GRanges(chr, IRanges(result$start_position,
                                          result$end_position),
                                       strand = "*",
                                       count=result$count))
                        
dTrack <- DataTrack(data, name="Count")

cdr_region <-data.frame(start = c(138369119, 100117325, 144345422),
                        end = c(140078203, 102627325, 148972761))


ht <- HighlightTrack(trackList=list(atrack, dTrack), 
                       start = cdr_region$start, 
                       end = cdr_region$end,
                       chromosome = chr,
                       genome = gen)
plotTracks(list(itrack, gtrack, ht), type = "b")
```


```{r TS Genes in Both Regression RF and Data mining}

chr <- 'chr7'
gen <- 'hg38'
result <- data.frame(regandjeremy.result, count = regandjeremy.count)
result <- result %>% filter(count == 2)

data_g <- with(result, GRanges(chr, IRanges(result$start_position,
                                            result$end_position),
                               result$strand,
                               group=result$hgnc_symbol,
                               id=result$hgnc_symbol))
atrack <- AnnotationTrack(data_g, name = "TS Genes in Both Regression RF and Data Mining",
                          groupAnnotation="group", fontcolor.group="darkblue", 
                          shape = "box", just.group = "left")
                            
gtrack <- GenomeAxisTrack()
  
itrack <- IdeogramTrack(genome = gen, chromosome = chr)

cdr_region <-data.frame(start = c(138369119, 100117325, 144345422),
                        end = c(140078203, 102627325, 148972761))


ht <- HighlightTrack(trackList=list(atrack), 
                       start = cdr_region$start, 
                       end = cdr_region$end,
                       chromosome = chr,
                       genome = gen)
plotTracks(list(itrack, gtrack, ht), )

```


```{r More }
regrf.complete.result <- data_retrieve(regrf.complete$Gene, mart.hs)
```


```{r}
chr <- 'chr7'
gen <- 'hg38'
result <- regrf.complete.result
result <- data.frame(result, pred_prob = regrf.complete$pred_prob)
result <- result[order(result$start_position), ]
  
data_g <- with(result, GRanges(chr, IRanges(result$start_position,
                                              result$end_position),
                               result$strand,
                               group=result$hgnc_symbol,
                               id=result$hgnc_symbol))
atrack <- AnnotationTrack(data_g, name = "Regression RF All Chr7 Genes")
                            
gtrack <- GenomeAxisTrack()
  
itrack <- IdeogramTrack(genome = gen, chromosome = chr)

data <- with(rank_data, GRanges(chr, IRanges(result$start_position,
                                             result$end_position),
                                strand = "*",
                                pred_prob=result$pred_prob))
                        
dTrack <- DataTrack(data, name="Predicted Probability", type = "l")
#dTrack_s <- DataTrack(data, name="Predicted Probability", type = "l")

cdr_region <-data.frame(start = c(138369119, 100117325, 144345422),
                        end = c(140078203, 102627325, 148972761))


ht <- HighlightTrack(trackList=list(dTrack), 
                       start = cdr_region$start, 
                       end = cdr_region$end,
                       chromosome = chr,
                       genome = gen)
plotTracks(list(itrack, gtrack, ht))

```


```{r}
regrf.complete.result <- data.frame(regrf.complete.result, 
                                    pred_prob = regrf.complete$pred_prob)
regrf.complete.result <- regrf.complete.result[order(regrf.complete.result$start_position), ]

cdr_zoomin_plot_w_pred_prob <- function(result, title, cdr_num) {
  
  cdr <-data.frame(start = c(100117325, 138369119, 144345422),
                   end = c(102627325, 140078203, 148972761))
  
  cdr_start <- cdr$start[cdr_num]
  cdr_end <- cdr$end[cdr_num]
  
  result <- result %>% filter((start_position >= cdr_start & end_position <= cdr_end) |
                              (start_position <= cdr_start & end_position >= cdr_start) |
                              (start_position <= cdr_end & end_position >= cdr_end) |
                              (start_position <= cdr_start & end_position >= cdr_end))
  
  if (nrow(result) == 0) {
    statement = paste("No TS Genes in", title)
    return (statement)
  }
  
  chr <- 'chr7'
  gen <- 'hg38'
  
  data_g <- with(result, GRanges(chr, IRanges(result$start_position,
                                              result$end_position),
                                 result$strand,
                                 group = result$hgnc_symbol,
                                 id=result$hgnc_symbol))
  atrack <- AnnotationTrack(data_g, name = title, 
                            groupAnnotation="group", fontcolor.group="darkblue", 
                            shape = "box", just.group = "left")
  
  gtrack <- GenomeAxisTrack()
  
  itrack <- IdeogramTrack(genome = gen, chromosome = chr)
  
  data<- with(rank_data, GRanges(chr, IRanges(result$start_position,
                                              result$end_position),
                                         strand = "*",
                                         pred_prob=result$pred_prob))
  
  dtrack <- DataTrack(data, name="Predicted Probability")
  
  if ((sum(result$start_position < cdr_start) + sum(result$end_position > cdr_end)) > 0) {
    ht <- HighlightTrack(trackList=list(atrack, dtrack), 
                         start = cdr_start, 
                         end = cdr_end,
                         chromosome = chr,
                         genome = gen)
    plotTracks(list(itrack, gtrack, ht), type = "l")
  } else {
    plotTracks(list(itrack, gtrack, atrack, dtrack), from = cdr_start - 3e5, to = cdr_end, type = "l")
  }
}
```


```{r zoom_in plot for regrf CDR1}
cdr_zoomin_plot_w_pred_prob(regrf.complete.result, "Regression RF CDR1 with Pred Prob", 1)
```


```{r zoom_in plot for regrf CDR2}
cdr_zoomin_plot_w_pred_prob(regrf.complete.result, "Regression RF CDR2 with Pred Prob", 2)
```


```{r zoom_in plot for regrf CDR3}
cdr_zoomin_plot_w_pred_prob(regrf.complete.result, "Regression RF CDR3 with Pred Prob", 3)
```


```{r}
regrf.complete
```


define a function that takes 2 list and 1 overlayed list with unique entries and return an array with 1s and 2s
```{r overlay function}
#define a function that takes 2 list and 1 overlayed list with unique entries and return an array with 1s and 2s

overlay_count <- function(overlay_list, list1, list2) {
  result <- c()
  
  #if an item in the overlayed list is in both list, then the entry will be 
  for (item in overlay_list) {
    if ((item %in% list1) & (item %in% list2)) {
      result <- append(result, 2)
    } else {
      result <- append(result, 1)
    }
  }
  return(result)
}

classandjeremy.ts <- unique(as.vector(c(classrf.ts, jeremy.result$hgnc_symbol)))
classandjeremy.count <- overlay_count(classandjeremy.ts, classrf.ts, jeremy.result$hgnc_symbol)

regandjeremy.ts <- unique(as.vector(c(regrf.ts, jeremy.result$hgnc_symbol)))
regandjeremy.count <- overlay_count(regandjeremy.ts, regrf.ts, jeremy.result$hgnc_symbol)
```


```{r, fig.width= 7, fig.height = 8}
chr <- 'chr7'
gen <- 'hg38'
#our two result dataframes are:
#exp.hits.result
#classrf.hits.result

#create the GRange object for ML result
data_classrf <- with(classrf.hits.result, GRanges(chr, IRanges(classrf.hits.result$start_position,
                                            classrf.hits.result$end_position),
                                            strand = "*",
                                            genome = gen,
                               #classrf.hits.result$strand,
                        count = classrf.hits.result$TS_freq))
#create a GRange object for experimental hits result
data_exp <- with(exp.hits.result, GRanges(chr, IRanges(exp.hits.result$start_position,
                                            exp.hits.result$end_position),
                                          strand = "*",
                                          genome = gen,
                               #exp.hits.result$strand,
                               count = exp.hits.result$z_comb))

atrack1 <- AnnotationTrack(data_classrf,name = "Genes")
atrack2 <-AnnotationTrack(data_exp,name = "Genes")
                            
gtrack <- GenomeAxisTrack()
  
itrack <- IdeogramTrack(genome = gen, chromosome = chr)

# data <- with(result, GRanges(chr, IRanges(result$start_position,
#                                           result$end_position),
#                                        strand = "*",
#                                        count=result$count))
                        
dtrack1 <- DataTrack(data_classrf,name = "classification rf score")
dtrack2 <- DataTrack(data_exp, name = "experimental combined score")

ot <- OverlayTrack(trackList=list(dtrack1, dtrack2))

cdr_region <-data.frame(start = c(138369119, 100117325, 144345422),
                        end = c(140078203, 102627325, 148972761))


ht1 <- HighlightTrack(trackList=list(atrack1, dtrack1), 
                       start = cdr_region$start, 
                       end = cdr_region$end,
                       chromosome = chr,
                       genome = gen)

ht2 <- HighlightTrack(trackList=list(atrack1, dtrack2), 
                       start = cdr_region$start, 
                       end = cdr_region$end,
                       chromosome = chr,
                       genome = gen)

ylims <- extendrange(range(c(values(dtrack1), values(dtrack2))))


plotTracks(list(itrack,gtrack,ht1,dtrack1), type = "p")
plotTracks(list(itrack,gtrack,ht2,dtrack2), type = "p") #ht1 contains all genes on chr7, so we still use it here

#plot side by side
plotTracks(list(itrack, gtrack, ot), #from = lim[1], to = lim[2], 
           ylim = ylims, type = c("smooth", "p"))
```

