{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Classification Train-test data processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Author: Shirley Zhou"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exclude non-coding gene from dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Since `biomaRt` does not function well, the lists of protein-coding gene in train and test set are obtained from Weihan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "chr7 = pd.read_csv(\"../data/Chr7_master_list.csv\").drop(['Unnamed: 0'], axis = 1)\n",
    "gwide = pd.read_csv(\"../data/Genomewide_master_list.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Aliases</th>\n",
       "      <th>description</th>\n",
       "      <th>other_designations</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>HSC.logCPM</th>\n",
       "      <th>Brummelkamp.hap1.GTS.ratio</th>\n",
       "      <th>Brummelkamp.hap1.GTS.threshold</th>\n",
       "      <th>Brummelkamp.kbm7.GTS.ratio</th>\n",
       "      <th>...</th>\n",
       "      <th>Sabbatini2017.THP.1</th>\n",
       "      <th>Sabbatini2017.THP.1.threshold</th>\n",
       "      <th>Wallace2016.DEseq2.Log2.Fold.Change</th>\n",
       "      <th>Wallace2016.DEseq2.P.Value</th>\n",
       "      <th>Wallace2016.threshold</th>\n",
       "      <th>Rep1_2.Mann.Whitney.p.value</th>\n",
       "      <th>average.phenotype.of.strongest.3.2</th>\n",
       "      <th>Unnamed: 83</th>\n",
       "      <th>Rep1_2.Mann.Whitney.p.value.1</th>\n",
       "      <th>average.phenotype.of.strongest.3.2.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>IGDCC3</td>\n",
       "      <td>9543</td>\n",
       "      <td>HsT18880, PUNC</td>\n",
       "      <td>immunoglobulin superfamily DCC subclass member 3</td>\n",
       "      <td>; immunoglobulin superfamily DCC subclass memb...</td>\n",
       "      <td>15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.503403</td>\n",
       "      <td>0</td>\n",
       "      <td>0.534752</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.130</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>2.320000e-07</td>\n",
       "      <td>-0.211542</td>\n",
       "      <td>1</td>\n",
       "      <td>0.692807</td>\n",
       "      <td>0.004401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>HSF2BP</td>\n",
       "      <td>11077</td>\n",
       "      <td>NaN</td>\n",
       "      <td>heat shock transcription factor 2 binding protein</td>\n",
       "      <td>; heat shock factor 2-binding protein</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.514247</td>\n",
       "      <td>0</td>\n",
       "      <td>0.521775</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.221</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>4.682615e-02</td>\n",
       "      <td>-0.006646</td>\n",
       "      <td>1</td>\n",
       "      <td>0.789091</td>\n",
       "      <td>-0.007323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ADGRA2</td>\n",
       "      <td>25960</td>\n",
       "      <td>GPR124, TEM5</td>\n",
       "      <td>adhesion G protein-coupled receptor A2</td>\n",
       "      <td>; adhesion G protein-coupled receptor A2|G-pro...</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>TKFC</td>\n",
       "      <td>26007</td>\n",
       "      <td>DAK, NET45</td>\n",
       "      <td>triokinase and FMN cyclase</td>\n",
       "      <td>; triokinase/FMN cyclase|ATP-dependent dihydro...</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>ERMARD</td>\n",
       "      <td>55780</td>\n",
       "      <td>C6orf70, PVNH6, dJ266L20.3</td>\n",
       "      <td>ER membrane associated RNA degradation</td>\n",
       "      <td>; endoplasmic reticulum membrane-associated RN...</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.070</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>4.053098e-01</td>\n",
       "      <td>0.010147</td>\n",
       "      <td>0</td>\n",
       "      <td>0.462927</td>\n",
       "      <td>-0.012278</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 86 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene  GeneID                     Aliases  \\\n",
       "0  IGDCC3    9543              HsT18880, PUNC   \n",
       "1  HSF2BP   11077                         NaN   \n",
       "2  ADGRA2   25960                GPR124, TEM5   \n",
       "3    TKFC   26007                  DAK, NET45   \n",
       "4  ERMARD   55780  C6orf70, PVNH6, dJ266L20.3   \n",
       "\n",
       "                                         description  \\\n",
       "0   immunoglobulin superfamily DCC subclass member 3   \n",
       "1  heat shock transcription factor 2 binding protein   \n",
       "2             adhesion G protein-coupled receptor A2   \n",
       "3                         triokinase and FMN cyclase   \n",
       "4             ER membrane associated RNA degradation   \n",
       "\n",
       "                                  other_designations chromosome  HSC.logCPM  \\\n",
       "0  ; immunoglobulin superfamily DCC subclass memb...         15         NaN   \n",
       "1              ; heat shock factor 2-binding protein         21         NaN   \n",
       "2  ; adhesion G protein-coupled receptor A2|G-pro...          8         NaN   \n",
       "3  ; triokinase/FMN cyclase|ATP-dependent dihydro...         11         NaN   \n",
       "4  ; endoplasmic reticulum membrane-associated RN...          6         NaN   \n",
       "\n",
       "   Brummelkamp.hap1.GTS.ratio  Brummelkamp.hap1.GTS.threshold  \\\n",
       "0                    0.503403                               0   \n",
       "1                    0.514247                               0   \n",
       "2                         NaN                               0   \n",
       "3                         NaN                               0   \n",
       "4                         NaN                               0   \n",
       "\n",
       "   Brummelkamp.kbm7.GTS.ratio  ...  Sabbatini2017.THP.1  \\\n",
       "0                    0.534752  ...               -0.130   \n",
       "1                    0.521775  ...               -0.221   \n",
       "2                         NaN  ...                  NaN   \n",
       "3                         NaN  ...                  NaN   \n",
       "4                         NaN  ...               -0.070   \n",
       "\n",
       "   Sabbatini2017.THP.1.threshold  Wallace2016.DEseq2.Log2.Fold.Change  \\\n",
       "0                              0                                  NaN   \n",
       "1                              0                                  NaN   \n",
       "2                              0                                  NaN   \n",
       "3                              0                                  NaN   \n",
       "4                              0                                  NaN   \n",
       "\n",
       "   Wallace2016.DEseq2.P.Value  Wallace2016.threshold  \\\n",
       "0                         NaN                      0   \n",
       "1                         NaN                      0   \n",
       "2                         NaN                      0   \n",
       "3                         NaN                      0   \n",
       "4                         NaN                      0   \n",
       "\n",
       "   Rep1_2.Mann.Whitney.p.value average.phenotype.of.strongest.3.2  \\\n",
       "0                 2.320000e-07                          -0.211542   \n",
       "1                 4.682615e-02                          -0.006646   \n",
       "2                          NaN                                NaN   \n",
       "3                          NaN                                NaN   \n",
       "4                 4.053098e-01                           0.010147   \n",
       "\n",
       "   Unnamed: 83  Rep1_2.Mann.Whitney.p.value.1  \\\n",
       "0            1                       0.692807   \n",
       "1            1                       0.789091   \n",
       "2            0                            NaN   \n",
       "3            0                            NaN   \n",
       "4            0                       0.462927   \n",
       "\n",
       "   average.phenotype.of.strongest.3.2.1  \n",
       "0                              0.004401  \n",
       "1                             -0.007323  \n",
       "2                                   NaN  \n",
       "3                                   NaN  \n",
       "4                             -0.012278  \n",
       "\n",
       "[5 rows x 86 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwide.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
